keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/ConstraintState.kt

```diff
@@ -0,0 +1,34 @@
+package com.netflix.spinnaker.keel.api
+
+import java.time.Duration
+import java.time.Instant
+
+data class ConstraintState(
+  val environmentUid: String,
+  val artifactVersion: String,
+  val constraintType: String,
+  val status: ConstraintStatus,
+  val createdAt: Instant = Instant.now(),
+  val judgedBy: String? = null,
+  val judgedAt: Instant? = null,
+  val comment: String? = null,
+  val json: String? = null
+) {
+  fun canPromote() = status.passes()
+
+  fun failed() = status.failed()
+
+  fun timedOut(timeout: Duration, now: Instant) =
+    createdAt.plus(timeout).isBefore(now)
+}
+
+enum class ConstraintStatus(private val passed: Boolean, private val failed: Boolean) {
+  PENDING(false, false),
+  PASS(true, false),
+  FAIL(false, true),
+  OVERRIDE_PASS(true, false),
+  OVERRIDE_FAIL(false, true);
```

Comment:
I find these a little confusing as they have the same values as `PASS` and `FAIL`, respectively. What's the indented use?

Reply:
Canary constraint fails, status -> FAIL
Human manually overrides the canary failure, status -> OVERRIDE_PASS